[AGW] mobilityD: add multi APN ip allocator

pshelar opened the pull request:
## Summary

Add support for Allocating IP address as per APN config.
This patch extends DHCP allocation as per APN vlan id.
It adds vlan_id parameter to ip_desc to keep vlan-id of 
the IP allocation. This is used during ip address release.
 
Signed-off-by: Pravin B Shelar <pbshelar@fb.com>




<!-- Enumerate changes you made and why you made them -->

## Test Plan

`make test`

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

xjtian commented on lte/gateway/python/magma/mobilityd/tests/test_multi_apn_ip_alloc.py:
```diff
@@ -0,0 +1,307 @@
+class StaticIPAllocationTests(unittest.TestCase):
+    """
+    Test class for the Mobilityd Static IP Allocator
+    """
```
Bad copypasta. This is the vlan allocator test. Rename the class too

pshelar replied:
fixed.

xjtian commented on lte/gateway/python/magma/mobilityd/tests/test_static_ip_alloc.py:
```diff
@@ -83,7 +83,7 @@ def clear_subs(cls):
 
 
-class StaticIPAllocationTests(unittest.TestCase):
+class MultiAPNIPAllocationTests(unittest.TestCase):
```
revert

pshelar replied:
done

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -60,3 +61,48 @@ def get_subscriber_ip(self, sid: str) -> ipaddress.ip_address:
+                        break
+
+                if vlan_id is not None:
+                    self.apn_vlan_map[apn_name] = str(int(vlan_id))
```
why aren't you sending the vlan ID down as an int in the first place and keeping it that way?

pshelar replied:
ok, I can change it to int.

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -21,6 +21,7 @@
 class SubscriberDbClient:
     def __init__(self, subscriberdb_rpc_stub):
         self.subscriber_client = subscriberdb_rpc_stub
+        self.apn_vlan_map = {}
```
why bother caching these in memory? how hot is this codepath going to be and will RPC overhead be significant enough? you're going to have to do some cache invalidation if you want to do this because these mappings could change.

pshelar replied:
I can remove cache, we can introduce it if needed.
I thought APN reconf would trigger service restart which invalidates cache.

xjtian commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -145,7 +145,8 @@ def release_ip(self, ip_desc: IPDesc, vlan: str = ""):
             vlan: vlan id of the APN.
```
update the docstring

xjtian commented on lte/gateway/python/magma/mobilityd/ip_allocator_base.py:
```diff
@@ -44,7 +44,7 @@ def list_allocated_ips(self, ipblock: ip_network) -> List[ip_address]:
 
     @abstractmethod
-    def alloc_ip_address(self, sid: str) -> IPDesc:
+    def alloc_ip_address(self, sid: str, vlan_id: str = "") -> IPDesc:
```
make the vlan ID a positional arg and refactor all the callsites instead

xjtian replied:
it doesn't look like you ever call `alloc_ip_address` without `vlan_id` specified, drop the default value

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -60,3 +61,48 @@ def get_subscriber_ip(self, sid: str) -> ipaddress.ip_address:
+            if data and data.non_3gpp and data.non_3gpp.apn_config:
+                vlan_id = None
+                for apn_config in data.non_3gpp.apn_config:
+                    if apn_config.service_selection == '*':
```
wildcard string should be a constant

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -60,3 +61,48 @@ def get_subscriber_ip(self, sid: str) -> ipaddress.ip_address:
+        TODO: Move this API to separate APN configuration service.
+        """
+        if self.subscriber_client is None:
+            return None
```
return type can't be str if you're returning None

pshelar committed:
mobilityD: add support for multi APN ip allocation.

Add support for Allocating IP address as per APN config.
This patch extends DHCP allocation as per APN vlan id.

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

xjtian commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -142,10 +142,11 @@ def release_ip(self, ip_desc: IPDesc, vlan: str = ""):
                 ip: IP assigned to this SID
                 ip_block: IP block of the IP address.
-            vlan: vlan id of the APN.
+                vlan_id: vlan id of the APN.
```
this is no longer a parameter

xjtian commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -165,7 +166,7 @@ def release_ip(self, ip_desc: IPDesc, vlan: str = ""):
         self._dhcp_client.stop()
 
-    def _alloc_ip_address_from_dhcp(self, mac: MacAddress, vlan: str) -> DHCPDescriptor:
+    def _alloc_ip_address_from_dhcp(self, mac: MacAddress, vlan: str = None) -> DHCPDescriptor:
```
A parameter type annotated as `str` cannot be `None`. Do you have any callsites that don't specify `vlan`? It doesn't look that way to me so remove the default value or refactor your callsites. If you need `None` as a valid input, this should be annotated as `Optional[str]`

xjtian commented on lte/gateway/python/magma/mobilityd/ip_allocator_multi_apn.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        return self._ip_allocator.list_allocated_ips(ipblock)
+
+    def alloc_ip_address(self, sid: str, vlan_id: str = None) -> IPDesc:
```
the default value violates the type annotation and is inconsistent with the parent interface

xjtian commented on lte/gateway/python/magma/mobilityd/ip_allocator_pool.py:
```diff
@@ -180,7 +180,7 @@ def list_allocated_ips(self, ipblock: ip_network) -> List[ip_address]:
         return res
 
-    def alloc_ip_address(self, sid: str) -> IPDesc:
+    def alloc_ip_address(self, sid: str, vlan: str = None) -> IPDesc:
```
this should match the parent interface

xjtian commented on lte/gateway/python/magma/mobilityd/ip_allocator_static.py:
```diff
@@ -62,13 +62,13 @@ def list_allocated_ips(self, ipblock: ip_network) -> List[ip_address]:
         return self._ip_allocator.list_allocated_ips(ipblock)
 
-    def alloc_ip_address(self, sid: str) -> IPDesc:
+    def alloc_ip_address(self, sid: str, vlan: str = None) -> IPDesc:
```
this should match the parent interface

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -30,33 +30,69 @@ def get_subscriber_ip(self, sid: str) -> ipaddress.ip_address:
+                    break
+
+            if ip_addr is not None:
+                return ipaddress.ip_address(ip_addr), str(vlan_id)
```
Why do we need the type coalescing on `vlan_id`?

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -30,33 +30,69 @@ def get_subscriber_ip(self, sid: str) -> ipaddress.ip_address:
+                    continue
+                if apn_config.service_selection == '*':
+                    ip_addr = apn_config.assigned_static_ip
+                    vlan_id = apn_config.vlan_id
```
I don't see this field on the proto on master, is this in a different changeset?

pshelar replied:
It is defined in this patch

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -30,33 +30,69 @@ def get_subscriber_ip(self, sid: str) -> ipaddress.ip_address:
+        TODO: Move this API to separate APN configuration service.
+        """
+        if self.subscriber_client is None:
+            return None
```
you're still returning `None` in a function that's return-annotated to never return `None`. So either return the empty string and refactor all your callers to accept that as a valid input (preferred), or update the return type and do None-checks in all your callers.

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -30,33 +30,69 @@ def get_subscriber_ip(self, sid: str) -> ipaddress.ip_address:
+            if ip_addr is not None:
+                return ipaddress.ip_address(ip_addr), str(vlan_id)
+
+        return None, None
```
You have to declare `Optional[str]` as the return if you ever return None

xjtian commented on lte/gateway/python/magma/mobilityd/ip_descriptor.py:
```diff
@@ -37,23 +37,31 @@ class IPDesc:
                      self.sid,
                      self.type)
+
+        if self.vlan_id is not None:
```
This conditional will always be true now

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -30,33 +30,69 @@ def get_subscriber_ip(self, sid: str) -> ipaddress.ip_address:
+                err.details())
+            return 0
+
+    def _find_ip_and_vlan(self, sid: str) -> (ipaddress.ip_address, int):
```
the first return should be `Optional[...]` unless you want to raise when the IP is not found

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -22,41 +24,77 @@ class SubscriberDbClient:
+                    break
+
+            if ip_addr is not None:
+                return ipaddress.ip_address(ip_addr), str(vlan_id)
```
get rid of the str conversion on vlan_id, your function is supposed to return an int

xjtian reviewed (approved):
Address the last issue before merging
